src/action/timeline.rs

```diff
@@ -188,52 +206,273 @@ impl super::Response for ChunkResponse {
 mod tests {
 
     use super::*;
+    use std::fs::{hard_link, create_dir, write, set_permissions, Permissions};
+    use std::ffi::OsStr;
+    use std::path::Path;
+    use std::os::unix::{ffi::OsStrExt, fs::{symlink, PermissionsExt}};
     use tempfile::tempdir;
     use crate::action::Request;
-    use std::os::unix::ffi::OsStrExt;
     use crate::gzchunked::GzChunkedDecoder;
 
+    fn handle_for_path<S>(session: &mut S, path: &Path) -> session::Result<()>
+    where
+        S: Session,
+    {
+        let path_bytes = Vec::from(path.as_os_str().as_bytes());
+        let args_proto = TimelineArgs { root: Some(path_bytes) };
+        let request = Request::from_proto(args_proto).unwrap();
+        handle(session, request)
+    }
+
+    fn entry_for_path(path: &Path) -> TimelineEntry {
+        let metadata = symlink_metadata(path).unwrap();
+        entry_from_metadata(&metadata, &PathBuf::from(path))
+    }
+
+    fn entries_from_session_response(session: &session::test::Fake) -> Vec<TimelineEntry> {
```

Comment:
These tests are now extremely complicated and super hard to follow. You have a lot of test utilites, that might be even more complicated than the code you are testing itself. How do you ensure that you don't have a bug in this code and it really tests what you want to test? This is one of the reasons why writing complex tests and a lot of test utilities is not a good idea (because you should probably write tests for them as well).

Tests should be extremely straightforward, just obvious to read. You can take a look at the tests of [the Python implementation](https://github.com/google/grr/blob/master/grr/client/grr_response_client/client_actions/timeline_test.py). They just use one two-line long helper for creating files and the rest is just a flat script without any branching and complicated loops (but actually, it should not even have these simple loops...).

Reply:
The Python implementation differs from my Rust one in lots of places, I don't think I can make my tests look as neat as Python ones.

For example, if we look at this specific function (`entries_from_session_response`), we can tell that most of it is just retrieving blocks from fake session transfer store and checking their digests. In Python, there's nothing like that because the FS walk and gzchunked serialization are implemented via coroutines, i.e. completely separated from session logic. I planned to do the exact same thing in the beginning, but you said that coroutines aren't idiomatic to Rust (yet) so I just stuck the whole thing to the session communication to avoid storing a lot of intermediate data.

I _guess_ I can do something about handler and directly test RecurseState instead of handle(), but the chunk unpacking here has to stay.

Or to be moved into fake session logic somehow, but is that how transfer store works in general and not in this specific action?

Reply:
Yeah, as the code is right now, it's hard to not have any helpers. Still, I think this `entries_from_session_response` is really all you need, and the rest should be just explicit assertions (instead of using `diff_entries`, loops and alike). Moreover, I don't really understand why you have some assertions in this method. All it should do is to parse all the data that you delivered to the transfer store and return it in form of `TimelineEntry`.

Reply:
We need these asserts because we would want to know if the chunk hashing has gone wrong for some reason. If I understand the whole process correctly, it would mean that the server won't be able to extract the blocks from the transfer store, which is pretty bad.

I also need `diff_entries` because if I remove it, I'll just have to copy its body into every test. I guess I'll just move the final assert out of it and only leave the transformation into diff, that would actually be nicer.

Reply:
> I also need diff_entries because if I remove it, I'll just have to copy its body into every test. I guess I'll just move the final assert out of it and only leave the transformation into diff, that would actually be nicer.

But why? You know what the output should be. You literally have a variable named `expected_entries` (which you should not—it even leads to a duplicated functionality, like you sort this vector before `diff_entries` and inside `diff_entries` in most cases and you likely did not notice this precisely because there is just too much logic even for simple tests).

The only test that *could* have some use of it is the one where you generate a deep directory structure. But as you said, the thing that you are testing there is the ability to handle deep directories, not whether the fields have the expected values. So I would argue that even there you don't need it.

Reply:
The `expected_entries` should be unordered because the directory listing order is undefined. I just couldn't find an actual unordered collection that could store protos so I have to compare sorted vectors.

> the thing that you are testing there is the ability to handle deep directories, not whether the fields have the expected values

Aren't those things equivalent?

And, I'm pretty sure that we discussed earlier that diff gives us more information than just assert_eq. Even if we only have 3-4 entries, it still saves us some time because it tells exactly which entries are deviating from expected ones.

Reply:
> The expected_entries should be unordered because the directory listing order is undefined. I just couldn't find an actual unordered collection that could store protos so I have to compare sorted vectors.

I don't see the connection, why ordering or not of the directory listing should affect your expectations? Because directory listing is undefined, in your test case you sort the output. And then it becomes "defined" (for the test sake). And you can make assertions on it.

> And, I'm pretty sure that we discussed earlier that diff gives us more information than just assert_eq. Even if we only have 3-4 entries, it still saves us some time because it tells exactly which entries are deviating from expected ones.

It gives you more information if you don't have explicit assertions on particular fields. If you have, then it's the same, except that you get a very confusing error message that you have to parse yourself and it requires a lot of logic.